[Draft PR] Add GPU-Accelerated NearestNeighbors on Spark

lijinf2 opened the pull request.

leewyang commented on src/spark_rapids_ml/core.py:
```diff
@@ -438,15 +454,20 @@ def _train_udf(pdf_iter: Iterator[pd.DataFrame]) -> pd.DataFrame:
+        if return_model == False:
+            return barrier_rdd.mapPartitions(lambda x: x)
+
+        ret = barrier_rdd.mapPartitions(lambda x: x).collect()[0]
```
I'd move `mapPartitions(lambda x: x)` into the common code above, since it's duplicated otherwise.

lijinf2 replied:
Done.

leewyang commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,344 @@
+        """
+        self.set_params(inputCol=value)
+        return self
+
```
Do we also need `setLabelCol()`?

lijinf2 replied:
Ah..., probably not. NN reused the labelCol to differentiate data and query vectors, but the standard NN application itself does not need the label column. I may be wrong.

leewyang commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,344 @@
+        Sets the value of `k`.
+        """
+        self.set_params(n_neighbors=value)
+        return self
```
This should be handled via parameter mapping, since mapping a Spark Param value to a differently named cuML param leads to weird behavior/inconsistencies.  So, add a `k` Spark ML Param and a `_param_mapping` entry to map Spark `k` to cuML `n_neighbors`.  (See `KMeans` for reference).

lijinf2 replied:
done.

leewyang commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,344 @@
+    ...         ([3.0, 3.0],),]
+    >>> topk = 2
+    >>> gpu_knn = NearestNeighbors().setInputCol("features").setK(topk)
+    >>> data_df = spark.SparkContext.parallelize(data).map(lambda row: (row,)).toDF(["features"])
```
Is the map to tuple required here?  Doesn't seem required in unit tests.

lijinf2 replied:
done.

leewyang reviewed (commented):
A couple comments

lijinf2 commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,344 @@
+        pass
+
+
+_set_pyspark_cuml_cls_param_attrs(NearestNeighbors, NearestNeighborsModel)
```
Yeah, removed the functional call and tested good.

lijinf2 commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,344 @@
+                handle=params["handle"],
+                n_neighbors=params[INIT_PARAMETERS_NAME]["n_neighbors"],
+                verbose=params[INIT_PARAMETERS_NAME]["verbose"],
+            )
```
Added MG class into the list of _cuml_cls. 
Edited the batch_size parameter into the test case test_nearest_neighbors.

lijinf2 commented on src/spark_rapids_ml/core.py:
```diff
@@ -358,7 +360,13 @@ def _enable_nccl(self) -> bool:
+
+    def _fit(
+        self, dataset: DataFrame, return_model: bool = True
+    ) -> Union["_CumlModel", RDD]:
```
Added a return_model function in the similar way as enable_nccl and enable_ucx.

leewyang reviewed (approved):
LGTM

eordentlich reviewed (commented):
In this example: https://spark.apache.org/docs/3.2.1/api/python/reference/api/pyspark.ml.feature.BucketedRandomProjectionLSH.html
there is an 'id' column.    We should probably support that.  

I'm actually not even sure that 'order' of the starting DF -> rdd is deterministic necessarily now that I think about it.

lijinf2 commented:
> In this example: https://spark.apache.org/docs/3.2.1/api/python/reference/api/pyspark.ml.feature.BucketedRandomProjectionLSH.html there is an 'id' column. We should probably support that.
> 
> I'm actually not even sure that 'order' of the starting DF -> rdd is deterministic necessarily now that I think about it.

Marked and will do (1) support id column (or neighboring columns) (2) ucx asyncio fix. 
Seem non-trivial so perhaps move to the next PRs to avoid this PR being too long? If so, will need approval again for this PR.

wbo4958 commented on src/spark_rapids_ml/core.py:
```diff
@@ -438,16 +457,23 @@ def _train_udf(pdf_iter: Iterator[pd.DataFrame]) -> pd.DataFrame:
+            if return_model == True:
+                if context.partitionId() == 0:
+                    yield pd.DataFrame(data=result)
+            else:
```
looks like this doesn't require context.barrier in line  439?

lijinf2 replied:
revised.

eordentlich commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,362 @@
+            # id mapping
+            id2row: Dict[int, int] = {}
+            count = 0
+            for r, _, item_rn in triplets:
```
Lot's of python for looping here and below (though some comprehension too).  We'll need to check how this does on GPU size scale data relative to the cuml part.

eordentlich replied:
Same goes for the Dict lookups.

lijinf2 replied:
created a ticket for future improvement.

eordentlich commented on tests/test_nearest_neighbors.py:
```diff
@@ -0,0 +1,144 @@
+        assert indices[4] == [7, 6]
+
+
+@pytest.mark.parametrize("feature_type", ["array"])
```
Does this mean only 'array' input type is currently supported?

lijinf2 replied:
Yeah, created a ticket to support multi-col df and vector df.

eordentlich commented:
The 

> > In this example: https://spark.apache.org/docs/3.2.1/api/python/reference/api/pyspark.ml.feature.BucketedRandomProjectionLSH.html there is an 'id' column. We should probably support that.
> > I'm actually not even sure that 'order' of the starting DF -> rdd is deterministic necessarily now that I think about it.
> 
> Marked and will do (1) support id column (or neighboring columns) (2) ucx asyncio fix. Seem non-trivial so perhaps move to the next PRs to avoid this PR being too long? If so, will need approval again for this PR.

For now let's at least return the input df's with the added zipped indices so that all the info is there and it is technically correct, though some post processing might be needed.   We can follow up and add a exact knn join api similar to the approxSimilarity join api of SparkML which lists all pairs of rows in dfA and dfB that are knn neighbors with the distance.

wbo4958 commented on src/spark_rapids_ml/core.py:
```diff
@@ -358,7 +360,15 @@ def _enable_nccl(self) -> bool:
+        """If _fit returns a model or a RDD"""
+        return True
+
+    def _fit(self, dataset: DataFrame) -> Union["_CumlModel", RDD]:
```
if you really want to return RDD. I'd like to suggest you move the code to a separate function to reuse some code instead of changing the "_fit"

lijinf2 replied:
Focusing on KNN functionality for now. 

Created a ticket to improve coding style in the future.

wbo4958 replied:
Could you move the common code in "_fit" into another function (eg, abc_func) to return the RDD. and in KNN please overwrite fit() which should call your abc_func instead of "_fit"

lijinf2 replied:
Copied the latest comment into the ticket for record. Will do this in the next PRs to avoid this one being too long.

eordentlich replied:
@wbo4958 @lijinf2 will do this or similar  in a followup pr.

wbo4958 commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,362 @@
+        self.data_df = self.df_zip_with_index(self.data_df)
+        return self
+
+    def kneighbors(self, query_df: DataFrame) -> Tuple[DataFrame, DataFrame]:
```
seems this is not pyspark ML way, the method neighbors should be in KNN model?

lijinf2 replied:
copied this comment to the last comment and the created ticket.

wbo4958 commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,362 @@
+        indices_df = res_df.select("indices")
+        return (distances_df, indices_df)
+
+    def df_zip_with_index(self, df: DataFrame) -> DataFrame:
```
Could you add some comments on this function to describe what's this API doing?

eordentlich replied:
Isn't internal.  Probably should be private with `_` prefix.

lijinf2 replied:
Revised.

wbo4958 commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,362 @@
+        )
+        zipped_rdd = df.rdd.zipWithIndex()
+        new_rdd = zipped_rdd.map(lambda row: [row[1]] + list(row[0]))
+        return new_rdd.toDF(schema=out_schema)
```
according to https://stackoverflow.com/a/56651510/1928940, this solution may be better, especially for spark-rapids.

eordentlich replied:
As written, does the added column actually need to be contiguous?  A mapping is constructed internally to the code and it doesn't assume anything about this column.

lijinf2 replied:
Not sure about the behavior of monotonically_increasing_id. Added a TODO comment and created a ticket.

eordentlich commented on src/spark_rapids_ml/common/cuml_context.py:
```diff
@@ -13,57 +13,142 @@
+            # the code occasionally run fail without setting the variables
+            my_ip = ips[self._rank]
+            my_ifname = CumlContext.get_ifname_from_ip(my_ip)
+            os.environ["UCX_TLS"] = "tcp,cuda_copy,cuda_ipc"
+            os.environ["UCXPY_IFNAME"] = my_ifname
```
We'll have to figure how to make this more flexible to take advantage of higher speed interconnects on multi-gpu nodes.

lijinf2 replied:
created a ticket.

lijinf2 commented on src/spark_rapids_ml/common/cuml_context.py:
```diff
@@ -13,57 +13,142 @@
+            ip, port = target_ip_ports[i]
+            ep = await ucx_worker.get_endpoint(ip, port)
+            eps[i] = ep
+        await asyncio.sleep(additional_timeout)
```
created a ticket.

lijinf2 commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,362 @@
+        res_df = res_df.sort("query_index")
+        distances_df = res_df.select("distances")
+        indices_df = res_df.select("indices")
+        return (distances_df, indices_df)
```
Revised code to return (query_df, item_df ,knn_df), where knn_df has columns (query_id, knn_indices, knn_distances).

lijinf2 commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,362 @@
+            messages = context.allGather(
+                message=json.dumps((rank, item_query_sizes, item_row_number))
+            )
+            triplets = [json.loads(msg) for msg in messages]
```
revised the variable name.

lijinf2 commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,362 @@
+            query_parts_to_ranks = []
+            for r, sizes, _ in triplets:
+                half_len = len(sizes) // 2
+                item_parts_to_ranks += [(r, size) for size in sizes[:half_len]]
```
revised.

lijinf2 commented:
@eordentlich @wbo4958 Addressed some comments and created tickets for the rest that seem to require further investigation. 
Please help review the PR again. Thanks.

eordentlich commented on src/spark_rapids_ml/knn.py:
```diff
@@ -90,15 +90,42 @@ class _NearestNeighborsCumlParams(_CumlParams, HasInputCol, HasLabelCol):
 
+    def setIdCol(self, value: str) -> "_NearestNeighborsCumlParams":
+        """
+        Sets the value of `id_col`. If not set, zipWithIndex will be used and the row order is not guaranted to be deterministic.
```
Is row order ever guaranteed to be deterministic? So not sure what is meant here.  Should just say an id column will be added with column name [...]  no need to mention zipWithIndex.   Also, somewhere should explain what an id column is used for (e.g., to specify nearest neighbor vectors by associated id value).

lijinf2 replied:
Revised the comment. Please check.

eordentlich commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,400 @@
+    """
+    Examples
+    --------
+    >>> from spark_rapids_ml.knn import NearestNeighbors
```
The example seems to need updating to match latest changes.

eordentlich commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,400 @@
+        ) = super()._pre_process_data(dataset)
+
+        if self.hasParam("id_col") and self.isDefined("id_col"):
+            label_name = self.getOrDefault("id_col")
```
Why label_name as variable name here?

lijinf2 replied:
Good catch. Revised the variable name to id_col_name.

eordentlich commented on src/spark_rapids_ml/knn.py:
```diff
@@ -0,0 +1,362 @@
+            import json
+
+            messages = context.allGather(
+                message=json.dumps((rank, item_query_sizes, item_row_number))
```
Any comment on this?

lijinf2 replied:
Good question. Created a ticket to investigate the performance of json dumps/loads and the potential to leverage ucx.

eordentlich commented:
Can you rebase?

lijinf2 commented:
> Can you rebase?

Done. Please check.

eordentlich commented:
👍
